pkg/controller/clusterstate/clusterstate_controller.go

```diff
@@ -0,0 +1,301 @@
+}
+
+// Reconcile ensures that a given ClusterState resource exists and reflects the state of cluster operators from its target cluster
+// +kubebuilder:rbac:groups=hive.openshift.io,resources=clusterstates;clusterstates/status,verbs=get;list;watch;create;update;patch;delete
+func (r *ReconcileClusterState) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	start := time.Now()
+	logger := r.logger.WithFields(log.Fields{
+		"controller":        controllerName,
+		"clusterDeployment": request.NamespacedName.String(),
+	})
+
+	// For logging, we need to see when the reconciliation loop starts and ends.
+	logger.Info("reconciling cluster deployment")
+	defer func() {
+		dur := time.Since(start)
+		hivemetrics.MetricControllerReconcileTime.WithLabelValues(controllerName).Observe(dur.Seconds())
+		logger.WithField("elapsed", dur).Info("reconcile complete")
+	}()
+
+	// Fetch the ClusterDeployment instance
+	cd := &hivev1.ClusterDeployment{}
+	err := r.Get(context.TODO(), request.NamespacedName, cd)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			logger.Debug("cluster deployment not found")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		logger.WithError(err).Error("Error getting cluster deployment")
+		return reconcile.Result{}, err
+	}
+	if !cd.DeletionTimestamp.IsZero() {
+		logger.Debug("ClusterDeployment resource has been deleted")
+		return reconcile.Result{}, nil
+	}
+	if !cd.Status.Installed || len(cd.Status.AdminKubeconfigSecret.Name) == 0 {
+		logger.Debug("ClusterDeployment is not yet ready")
+		return reconcile.Result{}, nil
+	}
+
+	// Fetch corresponding ClusterState instance
+	st := &hivev1.ClusterState{}
+	switch err = r.Get(context.TODO(), request.NamespacedName, st); {
+	case apierrors.IsNotFound(err):
+		logger.Info("Creating cluster state resource for cluster deployment")
+		st.Name = cd.Name
+		st.Namespace = cd.Namespace
+		if err = controllerutil.SetControllerReference(cd, st, r.scheme); err != nil {
+			logger.WithError(err).Error("error setting controller reference on cluster state")
+			return reconcile.Result{}, err
+		}
+		err = r.Create(context.TODO(), st)
+		if err != nil {
+			logger.WithError(err).Error("failed to create cluster state")
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	case err != nil:
+		logger.WithError(err).Error("Error getting cluster deployment")
+		return reconcile.Result{}, err
+	}
+	if !st.DeletionTimestamp.IsZero() {
+		logger.Debug("ClusterState resource has been deleted")
+		return reconcile.Result{}, nil
+	}
+	if st.Status.LastUpdated != nil && time.Since(st.Status.LastUpdated.Time) < statusUpdateInterval {
+		nextUpdateWait := statusUpdateInterval - time.Since(st.Status.LastUpdated.Time) + 60*time.Second
```

Comment:
Thinking more about it, no benefit... it will get reqeued anyway. I was thinking of the case where the amount of time elapsed is just shy of the interval, that it would maybe make little sense to RequeueAfter with a very small duration. But I guess it doesn't matter. Will remove.